### ISSUE TITLE
Due to [signal SIGSEGV: segmentation violation code] all management nodes crash in swarm mode.

### ISSUE THREAD

**Issue by pboers1988** [moby/moby#33613]
**Description**

On the latest version of docker 17.05.0-ce in swarm-mode whilst doing a rollback of a service all the management nodes in the cluster crash and are not able to restart.

**Steps to reproduce the issue:**
1.  docker service update name_of_service --rollback

**Describe the results you received:**
All management nodes crasht simultaneously and due to the fact that our cluster only consists of three nodes that are managers and workers at the same time the cluster failed catastrophically. 

**_After the failure it was not possible to restart the docker daemon, even after a node reboot. I had to downgrade to a previous version of docker_**

All nodes had similar output in /var/log/messages:

Jun  9 15:02:33 swarm-03-auto-prd dockerd: time="2017-06-09T15:02:33.664637652+02:00" level=info msg="3f651b434df29b4e [logterm: 212, index: 16787] sent MsgVote request to 61e553486e30b5fe at term 249" module=raft node.id=fr737euehhjgy3zthfn18jqat
Jun  9 15:02:33 swarm-03-auto-prd dockerd: time="2017-06-09T15:02:33.671321226+02:00" level=info msg="3f651b434df29b4e received MsgVoteResp from 61e553486e30b5fe at term 249" module=raft node.id=fr737euehhjgy3zthfn18jqat
Jun  9 15:02:33 swarm-03-auto-prd dockerd: time="2017-06-09T15:02:33.671360284+02:00" level=info msg="3f651b434df29b4e [quorum:2] has received 2 MsgVoteResp votes and 0 vote rejections" module=raft node.id=fr737euehhjgy3zthfn18jqat
Jun  9 15:02:33 swarm-03-auto-prd dockerd: time="2017-06-09T15:02:33.671378703+02:00" level=info msg="3f651b434df29b4e became leader at term 249" module=raft node.id=fr737euehhjgy3zthfn18jqat
Jun  9 15:02:33 swarm-03-auto-prd dockerd: time="2017-06-09T15:02:33.671390382+02:00" level=info msg="raft.node: 3f651b434df29b4e elected leader 3f651b434df29b4e at term 249" module=raft node.id=fr737euehhjgy3zthfn18jqat
Jun  9 15:02:33 swarm-03-auto-prd dockerd: panic: runtime error: invalid memory address or nil pointer dereference
**Jun  9 15:02:33 swarm-03-auto-prd dockerd: [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x14ad40f]**
Jun  9 15:02:33 swarm-03-auto-prd dockerd: goroutine 550 [running]:
Jun  9 15:02:33 swarm-03-auto-prd dockerd: panic(0x1726860, 0xc42000c050)
Jun  9 15:02:33 swarm-03-auto-prd dockerd: /usr/local/go/src/runtime/panic.go:500 +0x1a1
Jun  9 15:02:33 swarm-03-auto-prd dockerd: github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/orchestrator.IsTaskDirty(0xc4219f9200, 0xc420931680, 0xc4233a41a0)
Jun  9 15:02:33 swarm-03-auto-prd dockerd: /root/rpmbuild/BUILD/docker-engine/.gopath/src/github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/orchestrator/task.go:66 +0x3f
Jun  9 15:02:33 swarm-03-auto-prd dockerd: github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/orchestrator/update.(*Updater).isTaskDirty(0xc422f31c20, 0xc420931680, 0xc4230acf80)
Jun  9 15:02:33 swarm-03-auto-prd dockerd: /root/rpmbuild/BUILD/docker-engine/.gopath/src/github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/orchestrator/update/updater.go:524 +0x39
Jun  9 15:02:33 swarm-03-auto-prd dockerd: github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/orchestrator/update.(*Updater).isSlotDirty(0xc422f31c20, 0xc42314af00, 0x1, 0x1, 0xc4201c1e10)
Jun  9 15:02:33 swarm-03-auto-prd dockerd: /root/rpmbuild/BUILD/docker-engine/.gopath/src/github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/orchestrator/update/updater.go:528 +0x5c
Jun  9 15:02:33 swarm-03-auto-prd dockerd: github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/orchestrator/update.(*Updater).Run(0xc422f31c20, 0x7f4ccfa344a8, 0xc42093cc00, 0xc423488b80, 0x1, 0x1)
Jun  9 15:02:33 swarm-03-auto-prd dockerd: /root/rpmbuild/BUILD/docker-engine/.gopath/src/github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/orchestrator/update/updater.go:139 +0x113
Jun  9 15:02:33 swarm-03-auto-prd dockerd: github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/orchestrator/update.(*Supervisor).Update.func1(0xc422f31c20, 0x7f4ccfa344a8, 0xc42093cc00, 0xc423488b80, 0x1, 0x1, 0xc4230ab040, 0xc422b54a60, 0x19)
Jun  9 15:02:33 swarm-03-auto-prd dockerd: /root/rpmbuild/BUILD/docker-engine/.gopath/src/github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/orchestrator/update/updater.go:67 +0x64
Jun  9 15:02:33 swarm-03-auto-prd dockerd: created by github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/orchestrator/update.(*Supervisor).Update
Jun  9 15:02:33 swarm-03-auto-prd dockerd: /root/rpmbuild/BUILD/docker-engine/.gopath/src/github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/orchestrator/update/updater.go:73 +0x1fb
Jun  9 15:02:33 swarm-03-auto-prd systemd: docker.service: main process exited, code=exited, status=2/INVALIDARGUMENT
Jun  9 15:02:33 swarm-03-auto-prd systemd: Unit docker.service entered failed state.
Jun  9 15:02:33 swarm-03-auto-prd systemd: docker.service failed.
Jun  9 15:02:33 swarm-03-auto-prd systemd: docker.service holdoff time over, scheduling restart.
Jun  9 15:02:33 swarm-03-auto-prd systemd: Starting Docker Application Container Engine...
Jun  9 15:02:33 swarm-03-auto-prd dockerd: time="2017-06-09T15:02:33.940276621+02:00" level=info msg="libcontainerd: new containerd process, pid: 30803"
Jun  9 15:02:34 swarm-03-auto-prd dockerd: time="2017-06-09T15:02:34.941569569+02:00" level=warning msg="failed to rename /var/lib/docker/tmp for background deletion: %!s(<nil>). Deleting synchronously"
Jun  9 15:02:34 swarm-03-auto-prd dockerd: time="2017-06-09T15:02:34.944407027+02:00" level=info msg="[graphdriver] using prior storage driver: overlay"
Jun  9 15:02:34 swarm-03-auto-prd docker-volume-local-persist: Capabilities Called... Get Called...     Found portainer
Jun  9 15:02:34 swarm-03-auto-prd docker-volume-local-persist: Get Called...     Found portainer-cert
Jun  9 15:02:34 swarm-03-auto-prd dockerd: time="2017-06-09T15:02:34.956746792+02:00" level=info msg="Graph migration to content-addressability took 0.00 seconds"
Jun  9 15:02:34 swarm-03-auto-prd dockerd: time="2017-06-09T15:02:34.957287282+02:00" level=info msg="Loading containers: start."
Jun  9 15:02:35 swarm-03-auto-prd dockerd: time="2017-06-09T15:02:35.080343364+02:00" level=info msg="Default bridge (docker0) is assigned with an IP address 172.17.0.0/16. Daemon option --bip can be used to set a preferred IP address"
Jun  9 15:02:35 swarm-03-auto-prd dockerd: time="2017-06-09T15:02:35.124944396+02:00" level=info msg="Loading containers: done."
Jun  9 15:02:35 swarm-03-auto-prd dockerd: time="2017-06-09T15:02:35.157148419+02:00" level=info msg="Listening for local connections" addr="/var/run/docker/swarm/control.sock" module=node node.id=fr737euehhjgy3zthfn18jqat proto=unix
Jun  9 15:02:35 swarm-03-auto-prd dockerd: time="2017-06-09T15:02:35.157855367+02:00" level=info msg="Listening for connections" addr="145.145.64.119:2377" module=node node.id=fr737euehhjgy3zthfn18jqat proto=tcp
Jun  9 15:02:35 swarm-03-auto-prd dockerd: time="2017-06-09T15:02:35.310788586+02:00" level=warning msg="ignoring request to join cluster, because raft state already exists" module=node node.id=fr737euehhjgy3zthfn18jqat
Jun  9 15:02:35 swarm-03-auto-prd dockerd: time="2017-06-09T15:02:35.310858584+02:00" level=info msg="3f651b434df29b4e became follower at term 249" module=raft node.id=fr737euehhjgy3zthfn18jqat
Jun  9 15:02:35 swarm-03-auto-prd dockerd: time="2017-06-09T15:02:35.310875334+02:00" level=info msg="newRaft 3f651b434df29b4e [peers: [2914a2af982651d8,3f651b434df29b4e,61e553486e30b5fe], term: 249, commit: 16793, applied: 10000, lastindex: 16794, lastterm: 249]" module=raft node.id=fr737euehhjgy3zthfn18jqat
Jun  9 15:02:38 swarm-03-auto-prd dockerd: time="2017-06-09T15:02:38.941659728+02:00" level=info msg="3f651b434df29b4e [term: 249] received a MsgVote message with higher term from 61e553486e30b5fe [term: 251]" module=raft node.id=fr737euehhjgy3zthfn18jqat
Jun  9 15:02:38 swarm-03-auto-prd dockerd: time="2017-06-09T15:02:38.941754363+02:00" level=info msg="3f651b434df29b4e became follower at term 251" module=raft node.id=fr737euehhjgy3zthfn18jqat
Jun  9 15:02:38 swarm-03-auto-prd dockerd: time="2017-06-09T15:02:38.941776752+02:00" level=info msg="3f651b434df29b4e [logterm: 249, index: 16794, vote: 0] cast MsgVote for 61e553486e30b5fe [logterm: 249, index: 16794] at term 251" module=raft node.id=fr737euehhjgy3zthfn18jqat
Jun  9 15:02:38 swarm-03-auto-prd dockerd: time="2017-06-09T15:02:38.944272901+02:00" level=info msg="raft.node: 3f651b434df29b4e elected leader 61e553486e30b5fe at term 251" module=raft node.id=fr737euehhjgy3zthfn18jqat
Jun  9 15:02:40 swarm-03-auto-prd dockerd: time="2017-06-09T15:02:40.187210537+02:00" level=error msg="agent: session failed" error="session initiation timed out" module="node/agent" node.id=fr737euehhjgy3zthfn18jqat
Jun  9 15:02:42 swarm-03-auto-prd dockerd: time="2017-06-09T15:02:42.152122167+02:00" level=info msg="3f651b434df29b4e is starting a new election at term 251" module=raft node.id=fr737euehhjgy3zthfn18jqat
Jun  9 15:02:42 swarm-03-auto-prd dockerd: time="2017-06-09T15:02:42.152172620+02:00" level=info msg="3f651b434df29b4e became candidate at term 252" module=raft node.id=fr737euehhjgy3zthfn18jqat
Jun  9 15:02:42 swarm-03-auto-prd dockerd: time="2017-06-09T15:02:42.152186406+02:00" level=info msg="3f651b434df29b4e received MsgVoteResp from 3f651b434df29b4e at term 252" module=raft node.id=fr737euehhjgy3zthfn18jqat
Jun  9 15:02:42 swarm-03-auto-prd dockerd: time="2017-06-09T15:02:42.152198270+02:00" level=info msg="3f651b434df29b4e [logterm: 251, index: 16801] sent MsgVote request to 61e553486e30b5fe at term 252" module=raft node.id=fr737euehhjgy3zthfn18jqat
Jun  9 15:02:42 swarm-03-auto-prd dockerd: time="2017-06-09T15:02:42.152208395+02:00" level=info msg="3f651b434df29b4e [logterm: 251, index: 16801] sent MsgVote request to 2914a2af982651d8 at term 252" module=raft node.id=fr737euehhjgy3zthfn18jqat
Jun  9 15:02:42 swarm-03-auto-prd dockerd: time="2017-06-09T15:02:42.152218128+02:00" level=info msg="raft.node: 3f651b434df29b4e lost leader 61e553486e30b5fe at term 252" module=raft node.id=fr737euehhjgy3zthfn18jqat
Jun  9 15:02:42 swarm-03-auto-prd dockerd: time="2017-06-09T15:02:42.154724913+02:00" level=info msg="3f651b434df29b4e received MsgVoteResp from 61e553486e30b5fe at term 252" module=raft node.id=fr737euehhjgy3zthfn18jqat
Jun  9 15:02:42 swarm-03-auto-prd dockerd: time="2017-06-09T15:02:42.154765509+02:00" level=info msg="3f651b434df29b4e [quorum:2] has received 2 MsgVoteResp votes and 0 vote rejections" module=raft node.id=fr737euehhjgy3zthfn18jqat
Jun  9 15:02:42 swarm-03-auto-prd dockerd: time="2017-06-09T15:02:42.154782639+02:00" level=info msg="3f651b434df29b4e became leader at term 252" module=raft node.id=fr737euehhjgy3zthfn18jqat
Jun  9 15:02:42 swarm-03-auto-prd dockerd: time="2017-06-09T15:02:42.154793837+02:00" level=info msg="raft.node: 3f651b434df29b4e elected leader 3f651b434df29b4e at term 252" module=raft node.id=fr737euehhjgy3zthfn18jqat
Jun  9 15:02:42 swarm-03-auto-prd dockerd: panic: runtime error: invalid memory address or nil pointer dereference
Jun  9 15:02:42 swarm-03-auto-prd dockerd: [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x14ad40f]
Jun  9 15:02:42 swarm-03-auto-prd dockerd: goroutine 486 [running]:
Jun  9 15:02:42 swarm-03-auto-prd dockerd: panic(0x1726860, 0xc42000c050)
Jun  9 15:02:42 swarm-03-auto-prd dockerd: /usr/local/go/src/runtime/panic.go:500 +0x1a1
Jun  9 15:02:42 swarm-03-auto-prd dockerd: github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/orchestrator.IsTaskDirty(0xc420e225a0, 0xc422683680, 0xc422c40ff0)
Jun  9 15:02:42 swarm-03-auto-prd dockerd: /root/rpmbuild/BUILD/docker-engine/.gopath/src/github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/orchestrator/task.go:66 +0x3f
Jun  9 15:02:42 swarm-03-auto-prd dockerd: github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/orchestrator/update.(*Updater).isTaskDirty(0xc42341b8b0, 0xc422683680, 0x41ee68)
Jun  9 15:02:42 swarm-03-auto-prd dockerd: /root/rpmbuild/BUILD/docker-engine/.gopath/src/github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/orchestrator/update/updater.go:524 +0x39
Jun  9 15:02:42 swarm-03-auto-prd dockerd: github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/orchestrator/update.(*Updater).isSlotDirty(0xc42341b8b0, 0xc422c41160, 0x1, 0x1, 0x1)
Jun  9 15:02:42 swarm-03-auto-prd dockerd: /root/rpmbuild/BUILD/docker-engine/.gopath/src/github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/orchestrator/update/updater.go:528 +0x5c
Jun  9 15:02:42 swarm-03-auto-prd dockerd: github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/orchestrator/update.(*Updater).Run(0xc42341b8b0, 0x7f4ddd9c47e8, 0xc42092c3c0, 0xc4233ff680, 0x1, 0x1)
Jun  9 15:02:42 swarm-03-auto-prd dockerd: /root/rpmbuild/BUILD/docker-engine/.gopath/src/github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/orchestrator/update/updater.go:139 +0x113
Jun  9 15:02:42 swarm-03-auto-prd dockerd: github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/orchestrator/update.(*Supervisor).Update.func1(0xc42341b8b0, 0x7f4ddd9c47e8, 0xc42092c3c0, 0xc4233ff680, 0x1, 0x1, 0xc422fbdee0, 0xc422ec9b40, 0x19)
Jun  9 15:02:42 swarm-03-auto-prd dockerd: /root/rpmbuild/BUILD/docker-engine/.gopath/src/github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/orchestrator/update/updater.go:67 +0x64
Jun  9 15:02:42 swarm-03-auto-prd dockerd: created by github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/orchestrator/update.(*Supervisor).Update
Jun  9 15:02:42 swarm-03-auto-prd dockerd: /root/rpmbuild/BUILD/docker-engine/.gopath/src/github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/orchestrator/update/updater.go:73 +0x1fb
Jun  9 15:02:42 swarm-03-auto-prd systemd: docker.service: main process exited, code=exited, status=2/INVALIDARGUMENT
Jun  9 15:02:42 swarm-03-auto-prd systemd: Failed to start Docker Application Container Engine.
Jun  9 15:02:42 swarm-03-auto-prd systemd: Unit docker.service entered failed state.
Jun  9 15:02:42 swarm-03-auto-prd systemd: docker.service failed.
Jun  9 15:02:42 swarm-03-auto-prd systemd: docker.service holdoff time over, scheduling restart.
Jun  9 15:02:42 swarm-03-auto-prd systemd: Starting Docker Application Container Engine...

**Describe the results you expected:**

We expected a rollback of the service and not a catastrofic cluster failure :)

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
[pboers@swarm-01-auto-prd ~]$ docker version
Client:
 Version:      17.05.0-ce
 API version:  1.29
 Go version:   go1.7.5
 Git commit:   89658be
 Built:        Thu May  4 22:06:25 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.05.0-ce
 API version:  1.29 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   89658be
 Built:        Thu May  4 22:06:25 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
docker info
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 21
Server Version: 17.05.0-ce
Storage Driver: overlay
 Backing Filesystem: extfs
 Supports d_type: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local local-persist
 Network: bridge host macvlan null overlay
Swarm: active
 NodeID: zxk9layvgfupm364jobja9tvz
 Is Manager: true
 ClusterID: j1mm2qe05mpaebiwly21in590
 Managers: 3
 Nodes: 3
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: XXXXXXX
 Manager Addresses:
  1XXXXXX:2377
  XXXXXXX:2377
  XXXXXXX:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 9048e5e50717ea4497b757314bad98ea3763c145
runc version: 9c2d8d184e5da67c95d601382adf14862e4f2228
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-514.21.1.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 3.695GiB
Name: swarm-02-auto-prd.XXXXXXX
ID: 7DBR:PNQ4:LC7D:WESA:2T4C:NZ5V:YYA2:XXXXXXXXXXXXXXX
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Http Proxy: http://1XXXXXXXXXXX:3128/
Https Proxy: http://XXXXXXXXX:3128/
No Proxy: localhost,127.0.0.1,localaddress,.localdomain.com
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

**Comment by cpuguy83** [comment]
ping @aaronlehmann

**Comment by aaronlehmann** [comment]
Opened https://github.com/docker/swarmkit/pull/2233 to fix this.

**Comment by aaronlehmann** [comment]
Fixed by #33642

**Comment by thaJeztah** [comment]
Just checked with @aaronlehmann, and the fix for this issue is included in the 17.06 release
